diff --git a/arch/riscv/Kconfig b/arch/riscv/Kconfig
index fcbb81feb..fc48b458a 100644
--- a/arch/riscv/Kconfig
+++ b/arch/riscv/Kconfig
@@ -162,8 +162,8 @@ config MMU

 config PAGE_OFFSET
 	hex
-	default 0xC0000000 if 32BIT
-	default 0x80000000 if 64BIT && !MMU
+	default 0xC0000000 if 32BIT && MMU
+	default 0x80000000 if !MMU
 	default 0xff60000000000000 if 64BIT

 config KASAN_SHADOW_OFFSET
@@ -237,7 +237,6 @@ config ARCH_RV32I
 	select GENERIC_LIB_ASHRDI3
 	select GENERIC_LIB_LSHRDI3
 	select GENERIC_LIB_UCMPDI2
-	select MMU

 config ARCH_RV64I
 	bool "RV64I"

diff --git a/arch/riscv/include/uapi/asm/unistd.h b/arch/riscv/include/uapi/asm/unistd.h
index 73d7cdd2e..53f5410c1 100644
--- a/arch/riscv/include/uapi/asm/unistd.h
+++ b/arch/riscv/include/uapi/asm/unistd.h
@@ -15,9 +15,14 @@
  * along with this program.  If not, see <https://www.gnu.org/licenses/>.
  */

-#if defined(__LP64__) && !defined(__SYSCALL_COMPAT)
+#ifndef __SYSCALL_COMPAT
 #define __ARCH_WANT_NEW_STAT
 #define __ARCH_WANT_SET_GET_RLIMIT
+#endif
+
+#ifndef __LP64__
+#define __ARCH_WANT_STAT64
+#define __ARCH_WANT_TIME32_SYSCALLS
 #endif /* __LP64__ */

 #define __ARCH_WANT_SYS_CLONE3
-- 
2.38.1


